plugins/regrass.cpp

```diff
@@ -12,109 +22,423 @@ using namespace DFHack;
+    else if (mat == tiletype_material::STONE ||
+        mat == tiletype_material::LAVA_STONE ||
+        mat == tiletype_material::MINERAL)
+    {   // Muddy non-feature stone
+        for (auto blev : block->block_events)
+        {   // Remove mud spatter
+            if (blev->getType() != block_square_event_type::material_spatter)
+                continue;
+
+            auto &ms_ev = *(df::block_square_event_material_spatterst *)blev;
+            if (ms_ev.mat_type == builtin_mats::MUD)
+            {
+                ms_ev.amount[x][y] = 0;
+                DEBUG(log).print("Removed tile mud.\n");
+                break;
+            }
+        }
+    }
+
+    if (shape == tiletype_shape::FLOOR)
+    {   // Handle random variant, ashes
+        DEBUG(log).print("Tiletype to random grass floor.\n");
+        block->tiletype[x][y] = findRandomVariant((rand() & 1) ? tiletype::GrassLightFloor1 : tiletype::GrassDarkFloor1);
+    }
+    else
+    {
+        auto new_mat = (rand() & 1) ? tiletype_material::GRASS_LIGHT : tiletype_material::GRASS_DARK;
+        auto new_tt = findTileType(shape, new_mat, tiletype_variant::NONE, tiletype_special::NONE, nullptr);
+        DEBUG(log).print("Tiletype to %s.\n", ENUM_KEY_STR(tiletype, new_tt).c_str());
+        block->tiletype[x][y] = new_tt;
+    }
+
+    return 1;
+}
+
+int regrass_block(df::map_block *block, regrass_options options)
+{   // Regrass single block
+    int count = 0;
+    for (int y = 0; y < 16; y++)
+    {
+        for (int x = 0; x < 16; x++)
+            count += regrass_tile(block, x, y, options);
+    }
+
+    return count;
+}
+
+int regrass_map(regrass_options options)
+{   // Regrass entire map
+    int count = 0;
+    for (auto block : world->map.map_blocks)
+        count += regrass_block(block, options);
+
+    return count;
+}
+
+DFhackCExport command_result plugin_init(color_ostream &out, vector<PluginCommand> &commands)
+{
     commands.push_back(PluginCommand(
         "regrass",
         "Regrow surface grass and cavern moss.",
         df_regrass));
     return CR_OK;
 }
 
-command_result df_regrass(color_ostream &out, vector<string> & parameters) {
-    bool max = false;
-    if (!parameters.empty())
+command_result df_regrass(color_ostream &out, vector<string> &parameters)
+{
+    regrass_options options;
+
+    for (size_t i = 0; i < parameters.size(); i++)
     {
-        if(parameters[0] == "max")
-            max = true;
-        else
+        auto &s = parameters[i];
+        if (s == "m" || s == "max")
+            options.max_grass = true;
+        else if (s == "n" || s == "new")
+            options.new_grass = true;
+        else if (s == "a" || s == "ashes")
+            options.ashes = true;
+        else if (s == "u" || s == "mud")
+            options.mud = true;
+        else if (s == "p" || s == "here")
+            options.here = true;
+        else if (s == "b" || s == "block")
+            options.block = true;
+        else // "h", "help", invalid
```

Comment:
this provides a getopt-like interface, but is not getopt. Could you call out to Lua and use `argparse` instead? See `tailor.cpp` and `tailor.lua` for an example:

```cpp
    bool show_help = false;
    if (!Lua::CallLuaModuleFunction(out, "plugins.tailor", "parse_commandline", parameters, 1,
            [&](lua_State *L) {
                show_help = !lua_toboolean(L, -1);
            })) {
        return CR_FAILURE;
    }

    return show_help ? CR_WRONG_USAGE : CR_OK;
```

Reply:
Can Lua set the values of the `regrass_options` struct directly or do I need to workaround somehow?

Reply:
you can either pass the opts to Lua and modify the fields directly (like `prospector` and `blueprint` do) or you can call back into cpp from the commandline parser function according to the values on the commandline like `stockpiles` does.

After working with both styles, I prefer the latter, but both approaches work.